pennylane/operation.py

```diff
@@ -410,15 +413,26 @@ def parameters(self):
         """Current parameter values.
 
         Fixed parameters are returned as is, free parameters represented by
-        :class:`~.variable.Variable` instances are replaced by their
+        :class:`.VariableRef` instances are replaced by their
         current numerical value.
 
         Returns:
-            list[float]: parameter values
+            list[Any]: parameter values
         """
-        temp = list(_flatten(self.params))
-        temp_val = [self.check_domain(x.val, True) if isinstance(x, Variable) else x for x in temp]
-        return _unflatten(temp_val, self.params)[0]
+        # TODO profiling
+        def evaluate(p):
+            """Evaluate a single parameter."""
+            if isinstance(p, np.ndarray):
+                # object arrays may have VariableRefs inside them
+                if p.dtype == object:
+                    temp = np.array([x.val if isinstance(x, VariableRef) else x for x in p.flat])
+                    return temp.reshape(p.shape)
+                return p
+            if isinstance(p, VariableRef):
+                p = self.check_domain(p.val)
```

Comment:
I would expect a method `check_domain` only to check the domain of the parameter and not to return the given value

Reply:
It's for convenience, returning the value enables chaining the calls (kind of analogous to how `ostream::operator<<`  works in C++), and using the method in list comprehensions.